plugins/polio/models.py

```diff
@@ -262,10 +262,14 @@ def vaccine_names(self):
         campaign = self.campaign
 
         if campaign.separate_scopes_per_round:
-            scopes_with_orgunits = filter(lambda s: len(s.group.org_units.all()) > 0, self.scopes.all())
+            scopes_with_orgunits = filter(
+                lambda s: len(s.group.org_units.all()) > 0 and s.vaccine is not None, self.scopes.all()
+            )
             return ", ".join(scope.vaccine for scope in scopes_with_orgunits)
         else:
-            scopes_with_orgunits = filter(lambda s: len(s.group.org_units.all()) > 0, self.campaign.scopes.all())
+            scopes_with_orgunits = filter(
+                lambda s: len(s.group.org_units.all()) > 0 and s.vaccine is not None, self.campaign.scopes.all()
```

Comment:
```python
self.campaign.scopes
    .annotate(org_unit_count=Count("group__org_units"))
    .filter(Q(org_unit_count__gt=0) &Q( vaccine__isnull=False))
```

I don't know if it will be much more performant, but if I remember correctly, using list methods on the whole queryset is not optimal

Reply:
maybe a backend guru can weight in

Reply:
I know the code was already like that, but I think @quang-le is right and his suggestion more readable.

I (personally) don't like [the built-in Python `filter()`](https://docs.python.org/3/library/functions.html#filter) because:

- I have to think about what is happening every time I see it
- it's easily confused with Django `queryset.filter`

As far I understand, inside this `filter()`:

- `s.group.org_units.all()` will be triggered
- for every item in  `self.campaign.scopes.all()`

A proper unit test of `vaccine_names()` would have allowed to:

-  confirm this with absolute certainty
- refactor with confidence (knowing the expected output)

but unfortunately there isn't any…